hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala

```diff
@@ -444,23 +441,41 @@ abstract class RegistryFunctions {
       case TString => classTag[String]
       case TArray(TInt32) => classTag[IndexedSeq[Int]]
       case TArray(TFloat64) => classTag[IndexedSeq[Double]]
-      case TArray(TString) => classTag[IndexedSeq[String]]
+      case TArray(TString) => classTag[Array[String]]
       case TSet(TString) => classTag[Set[String]]
       case TDict(TString, TString) => classTag[Map[String, String]]
       case TCall => classTag[Int]
       case t => PrimitiveTypeToIRIntermediateClassTag(t)
     }
 
-    def wrap(cb: EmitCodeBuilder, r: Value[Region], code: SCode): Code[_] = code.st match {
+    def wrap(cb: EmitCodeBuilder, r: Value[Region], code: SCode): Code[_] = code.st.virtualType match {
       case t if t.isPrimitive => SType.extractPrimCode(cb, code)
-      case call: SCall => code.asCall.loadCanonicalRepresentation(cb)
+      case TCall => code.asCall.loadCanonicalRepresentation(cb)
+      case TArray(TString) => code.st match {
+        case _: SJavaArrayString => code.asInstanceOf[SJavaArrayStringCode].array
+        case _ =>
+          val sv = code.asIndexable.memoize(cb, "scode_array_string")
+          val arr = cb.newLocal[Array[String]]("scode_array_string", Code.newArray[String](sv.loadLength()))
+          sv.foreach(cb) { case (cb, idx, elt) =>
```

Comment:
why do you need this new `foreach` method rather than `foreachDefined` here?

Reply:
ah, that will work.